### PR TITLE
feat: implement delete user by email - management API 

### PR DESCRIPTION
**Pull request by danielgent**
This looks working now!  We have CRUD  

[Screencast from 2023-09-08 12-39-43.webm](https://github.com/sesamyab/auth/assets/8496063/3fc3d568-9dbc-4710-af2c-ed5051e1c7a1)

### CODE REVIEW

File: src/routes/tsoa/users-mgmt.ts [hunk 1]
```diff
@@ -53,6 +54,35 @@ export class UsersMgmtController extends Controller {
...
+
+    if (!dbUser) {
+      throw new NotFoundError();
+    }
+
+    const user = env.userFactory.getInstanceByName(
+      getId(tenantId, dbUser.email),
+    );
+
+    return user.delete.mutate();
```
Comment (danielgent) [inline src/routes/tsoa/users-mgmt.ts:83]: I'm not sure what magic is happening with `user.delete.mutate()` but the un-minified error doesn't indicate a whole lot

If I do `return user.getProfile.query()` then this returns the profile... So the user is in a durable object?

File: src/routes/tsoa/users.ts [hunk 2]
```diff
@@ -174,7 +174,7 @@ export class UsersController extends Controller {
   }
 
   @Delete("{userId}")
-  @SuccessResponse(201, "Created")
+  @SuccessResponse(201, "Delete")
```
Comment (danielgent) [inline src/routes/tsoa/users.ts:177]: Seems like this should be so :smile:

File: src/routes/tsoa/users-mgmt.ts [hunk 3]
```diff
@@ -37,7 +38,7 @@ export class UsersMgmtController extends Controller {
       .selectFrom("users")
       .where("users.tenantId", "=", tenantId)
       .where("users.id", "=", userId)
-      .selectAll()
+      .select("users.email")
```
Comment (danielgent) [inline src/routes/tsoa/users-mgmt.ts:41]: As mentioned in another PR review, no need to select all the fields

**Comment by danielgent** [comment]
I have no idea what's happening here @markusahlstrand !  There are libraries on top of libraries of tech that I don't understand.  I'll watch videos on cloudflare and play around with my demo project for a bit

I want to communicate with Durable Objects the standard way to try and understand what you've done

File: src/routes/tsoa/users-mgmt.ts (hunk 1 above)
Comment (markusahlstrand) [reply src/routes/tsoa/users-mgmt.ts:83]: Yes, so we're deleting the user in the DO and it's basically clearing the storage and then that removes the row from the database.

File: src/routes/tsoa/users.ts (hunk 2 above)
Comment (markusahlstrand) [reply src/routes/tsoa/users.ts:177]: A 201 is the http code for created, so think it should be a 200 or a 204?

File: src/models/User.ts [hunk 4]
```diff
@@ -293,21 +293,24 @@ export const userRouter = router({
 
       return profile;
     }),
-  delete: publicProcedure.mutation(async ({ ctx }) => {
-    const profile = await getProfile(ctx.env.storage);
+  delete: publicProcedure.mutation(
+    async ({ ctx }: { ctx: ContextWithBody<Env> }) => {
```
Comment (markusahlstrand) [inline src/models/User.ts:297]: Think this type is incorrect. This is the type of context you get in tsoa, but not in tRPC. In tRPC we also have context that contains an env, but it's different. If I remember correctly the storage is at ctx.state.storage.. Think typescript should give you the right type here?

**Review by markusahlstrand — COMMENTED** [review]
Think it's close but the types are not correct. It's different context in tRPC vs TSOA

File: src/routes/tsoa/users.ts (hunk 2 above)
Comment (danielgent) [reply src/routes/tsoa/users.ts:177]: I'll keep it simple as a 200 :+1:

File: src/models/User.ts (hunk 4 above)
Comment (danielgent) [reply src/models/User.ts:297]: I'll revert this now *but* the next line is what is blowing up. `ctx.env.storage` is undefined.  Looks to me like `storage` doesn't exist anywhere

Comment (danielgent) [reply src/models/User.ts:297]: There's no types being returned from the `ctx` here

File: src/models/User.ts [hunk 5]
```diff
@@ -294,7 +294,7 @@ export const userRouter = router({
       return profile;
     }),
   delete: publicProcedure.mutation(async ({ ctx }) => {
-    const profile = await getProfile(ctx.env.storage);
+    const profile = await getProfile(ctx.state.storage);
```
Comment (danielgent) [inline src/models/User.ts:297]: I've done this fix @markusahlstrand . It's a success in that we get another error

image


:upside_down_face: 

I'll add logs now

Comment (danielgent) [reply src/models/User.ts:298]: Nope. I think it's regressed. I'm not even hitting the first console log now

Comment (danielgent) [reply src/models/User.ts:298]: Ahh, I think the console logs from here don't make it out, that's why I couldn't just log out `ctx`

so we're flying even more blind than normal

The only way to work out what is throwing an error is to comment out lines

**Comment by danielgent** [comment]
Hmmmm, a step back!

Now the authentication step is failing because the shape of the clients has changed

image

File: src/types/Client.ts [hunk 6]
```diff
@@ -64,7 +64,8 @@ export const BaseClientSchema = z.object({
   tenantId: z.string(),
   clientSecret: z.string(),
   tenant: z.object({
-    audience: z.string(),
+    // this should fix /oauth/token issue?
+    audience: z.string().optional(),
```
Comment (danielgent) [inline src/types/Client.ts:68]: as done on https://github.com/sesamyab/auth/pull/121

**Review by markusahlstrand — APPROVED** [review]
Looks good! I'll fix up the audience in a separate PR.

**Pull request merged**